Proposal for Knative Eventing Spec Conformance Test Plan (#58)
knative/specs: specs/eventing/test-plan/broker-lifecycle-conformance.md

>> The annotation eventing.knative.dev/broker.class SHOULD be used to select a particular implementation of a Broker, if multiple implementations are available. It is RECOMMENDED to default the eventing.knative.dev/broker.class field on creation if it is unpopulated. Once created, the eventing.knative.dev/broker.class annotation and the spec.config field MUST be immutable; the Broker MUST be deleted and re-created to change the implementation class or spec.config. This pattern is chosen to make it clear that changing the implementation class or spec.config is not an atomic operation and that any implementation would be likely to result in event loss during the transition.

# Testing Broker Lifecycle Conformance: 

> We are going to be testing the previous paragraphs coming from the Knative Eventing Spec. To do this we will be creating a broker checking its immutable properties, checking its Ready Status and then creating a Trigger that links to it by making a reference. We will also checking the Trigger Status, as it depends on the Broker to be ready to work correclty. We will be also checking that the broker is addresable by looking at the status conditions fields. Because this is a Control Plane test, we are not going to be sending Events to these components. 

Reviewer: ```suggestion
We are going to be testing the previous paragraphs coming from the Knative Eventing Spec. To do this we will be creating a broker, checking its immutable properties, checking its Ready status and then creating a Trigger that links to it by making a reference. We will also checking the Trigger Status, as it depends on the Broker to be ready to work correctly. We will be also checking that the broker is addressable by looking at the status conditions fields. Because this is a Control Plane test, we are not going to be sending Events to these components. 
```

> We are going to be testing the previous paragraphs coming from the Knative Eventing Spec. To do this we will be creating a broker, checking its immutable properties, checking its Ready status and then creating a Trigger that links to it by making a reference. We will also checking the Trigger Status, as it depends on the Broker to be ready to work correctly. We will be also checking that the broker is addressable by looking at the status conditions fields. Because this is a Control Plane test, we are not going to be sending Events to these components. 

Reviewer 2: nit: Trigger Status -> Trigger status

Author: @lionelvillard done!

> You can find the resources for running these tests inside the `control-plane/broker-lifecycle/` directory. 
> - A broker resource: `control-plane/broker-lifecycle/broker.yaml`
> - A trigger resource that reference the broker: `control-plane/broker-lifecycle/trigger.yaml` 
> - A trigger resource that doesn't reference the broker: `control-plane/broker-lifecycle/trigger-no-broker.yaml`

Reviewer: ```suggestion
You can find the resources for running these tests inside the [control-plane/broker-lifecycle/](specs/eventing/test-plan/control-plane/broker-lifecycle/) directory. 
- A [broker resource](specs/eventing/test-plan/control-plane/broker-lifecycle/broker.yaml)
- A [trigger resource that references the broker](specs/eventing/test-plan/control-plane/broker-lifecycle/trigger.yaml)
- A [trigger resource that doesn't reference the broker](specs/eventing/test-plan/control-plane/broker-lifecycle/trigger-no-broker.yaml)
```

## [Pre] Creating a Broker 

```
kubectl apply -f control-plane/broker-lifecycle/broker.yaml
```

## [Test] Immutability

Check for default annotations, this should return the name of the selected implementation: 

```

> kubectl get broker conformance-broker -ojson | jq '.metadata.annotations["eventing.knative.dev/broker.class"]'

Reviewer: WDYT about using `-o jsonpath`? We might be able to get rid of the `jq` dependency

```suggestion
kubectl get broker conformance-broker -o jsonpath='{.metadata.annotations.eventing\.knative\.dev/broker\.class}'
```

Author: @benmoss yeah.. I was thinking about that.. but I decided on `jq` as it can do more advanced things than checking fields.. but it might have been an early optimization.. so let's remove `jq` for now until we really need it.

```

### [Output]

```
{
  "test": "control-plane/broker-lifecycle/broker-readyness"
  "output": {
  	"brokerImplementation": "<BROKER IMPLEMENTATION>",
	"expectedType": "Ready",
	"expectedStatus": "True"
  }
}
```

## [Test] Broker is Addresable

Running the following command should return a URL

```
kubectl get broker conformance-broker -ojson | jq .status.address.url
```

### [Output]

```
{
  "test": "control-plane/broker-lifecycle/broker-addressable"
  "output": {
  	"brokerImplementation": "",
	"obtainedURL": "<BROKER URL>",
  }
}
```

## [Pre] Create Trigger for Broker

Create a trigger that points to the broker:

```
kubectl apply -f control-plane/broker-lifecycle/trigger.yaml
```

## [Test] Broker Reference in Trigger

Check that the `Trigger` is making a reference to the `Broker`, this should return the name of the broker.

```
kubectl get trigger conformance-trigger -ojson | jq '.spec.broker'
```

### [Output]

```
{
  "test": "control-plane/broker-lifecycle/broker-reference-in-trigger"
  "output": {
  	"brokerImplementation": "<BROKER IMPLEMENTATION>",
	"expectedReference": "conformance-broker"
  }
}

> 

Reviewer 2: closing ``` is missing

Author: done!

## [Test] Trigger for Broker Readyness

Check for condition type `Ready` with status `True`: 

```
kubectl get trigger conformance-trigger -ojson | jq '.status.conditions[] |select(.type == "Ready")'
```

### [Output]

```
{
  "test": "control-plane/broker-lifecycle/trigger-for-broker-readyness"
  "output": {
  	"brokerImplementation": "<BROKER IMPLEMENTATION>",
	"expectedType": "Ready",
	"expectedStatus": "True"
  }
}
```

# Clean up & Congratulations

Make sure that you clean up all resources created in these tests by running: 

```
kubectl delete -f control-plane/broker-lifecycle/
```

> Congratulations you have tested the **Broker Lifecycle Conformance** :metal: !

Reviewer 2: :metal: does not render properly. Try 🤘 ?

Reviewer 2: might be because of the missing ```

Author: does it work now?

Reviewer 2: yes!

>> The annotation eventing.knative.dev/broker.class SHOULD be used to select a particular implementation of a Broker, if multiple implementations are available. It is RECOMMENDED to default the eventing.knative.dev/broker.class field on creation if it is unpopulated. Once created, the eventing.knative.dev/broker.class annotation and the spec.config field MUST be immutable; the Broker MUST be deleted and re-created to change the implementation class or spec.config. This pattern is chosen to make it clear that changing the implementation class or spec.config is not an atomic operation and that any implementation would be likely to result in event loss during the transition.

# Testing Broker Lifecycle Conformance: 

> We are going to be testing the previous paragraphs coming from the Knative Eventing Spec. To do this we will be creating a broker, checking its immutable properties, checking its Ready status and then creating a Trigger that links to it by making a reference. We will also checking the Trigger status, as it depends on the Broker to be ready to work correctly. We will be also checking that the broker is addressable by looking at the status conditions fields. Because this is a Control Plane test, we are not going to be sending Events to these components. 

Reviewer 3: ```suggestion
We are going to be testing the previous paragraphs coming from the Knative Eventing Spec. To do this we will be creating a Broker, checking its immutable properties, checking its Ready status and then creating a Trigger that links to it by making a reference. We will also checking the Trigger status, as it depends on the Broker to be ready to work correctly. We will be also checking that the Broker is addressable by looking at the status conditions fields. Because this is a Control Plane test, we are not going to be sending Events to these components. 
```

> You can find the resources for running these tests inside the [control-plane/broker-lifecycle/](specs/eventing/test-plan/control-plane/broker-lifecycle/) directory. 

Reviewer 2: this link is wrong (just remove `specs/eventing/test-plan/` prefix) as well as the 3 links below

- A [broker resource](specs/eventing/test-plan/control-plane/broker-lifecycle/broker.yaml)

> - A [trigger resource that references the broker](specs/eventing/test-plan/control-plane/broker-lifecycle/trigger.yaml)
> - A [trigger resource that doesn't reference the broker](specs/eventing/test-plan/control-plane/broker-lifecycle/trigger-no-broker.yaml)
> 

Reviewer 4: ```suggestion
- A [Trigger resource that references the Broker](specs/eventing/test-plan/control-plane/broker-lifecycle/trigger.yaml)
- A [Trigger resource that doesn't reference the Broker](specs/eventing/test-plan/control-plane/broker-lifecycle/trigger-no-broker.yaml)

```

```

Try to patch the annotation: `eventing.knative.dev/broker.class` to see if the resource mutates: 

```
kubectl patch broker conformance-broker --type merge -p '{"metadata":{"annotations":{"eventing.knative.dev/broker.class":"mutable"}}}'
```

You should get the following error: 
```
Error from server (BadRequest): admission webhook "validation.webhook.eventing.knative.dev" denied the request: validation failed: Immutable fields changed (-old +new): annotations
{string}:
	-: "MTChannelBasedBroker"
	+: "mutable"
```

### [Output]

```
{
  "test": "control-plane/broker-lifecycle/immutability-1"
  "output": {
    	"brokerImplementation": "<BROKER IMPLEMENTATION>",
	"expectedError": "<EXPECTED ERROR>"
  }
}
```

Try to mutate the `.spec.config` to see if the resource mutates: 

```
kubectl patch broker conformance-broker --type merge -p '{"spec":{"config":{"apiVersion":"v1"}}}'
```

### [Output]

```
{
  "test": "control-plane/broker-lifecycle/immutability-2"
  "output": {
  	"brokerImplementation": "<BROKER IMPLEMENTATION>",
	"expectedError": "<EXPECTED ERROR>"
  }
}
```

> ## [Test] Broker Readyness 

Reviewer 4: ```suggestion
## [Test] Broker Readiness 
```

Check for condition type `Ready` with status `True`: 

```
 kubectl get broker conformance-broker -ojsonpath="{.status.conditions[?(@.type == \"Ready\")].status}"
```

### [Output]

```
{

>   "test": "control-plane/broker-lifecycle/broker-readyness"

Reviewer 4: ```suggestion
  "test": "control-plane/broker-lifecycle/broker-readiness"
```

  "output": {
  	"brokerImplementation": "<BROKER IMPLEMENTATION>",
	"expectedType": "Ready",
	"expectedStatus": "True"
  }
}
```

## [Test] Broker is Addresable

Running the following command should return a URL

```
kubectl get broker conformance-broker -ojsonpath="{.status.address.url}"
```

### [Output]

```
{
  "test": "control-plane/broker-lifecycle/broker-addressable"
  "output": {
  	"brokerImplementation": "",
	"obtainedURL": "<BROKER URL>",
  }
}
```

## [Pre] Create Trigger for Broker

> Create a trigger that points to the broker:

Reviewer 4: Same here

> Create a trigger that points to the broker:

Reviewer 4: ```suggestion
Create a Trigger that points to the Broker:
```

```
kubectl apply -f control-plane/broker-lifecycle/trigger.yaml
```

## [Test] Broker Reference in Trigger

> Check that the `Trigger` is making a reference to the `Broker`, this should return the name of the broker.

Reviewer 4: And this convention is good, with the "`" and starting with upper-case, but we should have it consistent in all the file

```
kubectl get trigger conformance-trigger -ojsonpath="{.spec.broker}"
```

### [Output]

```
{
  "test": "control-plane/broker-lifecycle/broker-reference-in-trigger"
  "output": {
  	"brokerImplementation": "<BROKER IMPLEMENTATION>",
	"expectedReference": "conformance-broker"
  }
}
```

> ## [Test] Trigger for Broker Readyness

Reviewer 4: ```suggestion
## [Test] Trigger for Broker Readiness
```

Check for condition type `Ready` with status `True`: 

```

> kubectl get trigger conformance-trigger -ojson | jq '.status.conditions[] |select(.type == "Ready")'

Reviewer 2: instead of `jq`, what about using `-ojsonpath` ? The benefit is to remove the dependency on `jq`

Author: yes.. I think that is a good idea.. can you propose the changes?

Reviewer 2: `kubectl get trigger conformance-trigger -ojsonpath="{.status.conditions[?(@.type == \"Ready\")].status}"`

Output:

`True`

Author: @lionelvillard yeah.. I got convinced now.. let's remove `jq` from here :) I will do the other changes early on Monday.

Author: @lionelvillard `jq` removed now.

```

### [Output]

```
{

>   "test": "control-plane/broker-lifecycle/trigger-for-broker-readyness"

Reviewer 4: ```suggestion
  "test": "control-plane/broker-lifecycle/trigger-for-broker-readiness"
```
